src/main/java/frc/robot/systems/FSMSystem.java

```diff
@@ -1,132 +1,283 @@
+            case EXTENDED:
+                handleExtendedState(input);
+                break;
+                
+            case INTAKING_EXTENDED:
+                handleIntakingExtendedState(input);
+                break;
+                
+            case INTAKING_RETRACTED:
+                handleIntakingRetractedState(input);
+                break;
+    
+            case SHOOT_EXTENDED:
+                handleShootExtendedState(input);
+                break;
+    
+            case SHOOT_RETRACTED:
+                handleShootRetractedState(input);
+                break;
+    
+            case RETRACTED:
+                handleRetractedState(input);
+                break;
+    
+            default:
+                throw new IllegalStateException("Invalid state: " + currentState.toString());
+        }
 
-	/* ======================== Private methods ======================== */
-	/**
-	 * Decide the next state to transition to. This is a function of the inputs
-	 * and the current state of this FSM. This method should not have any side
-	 * effects on outputs. In other words, this method should only read or get
-	 * values to decide what state to go to.
-	 * @param input Global TeleopInput if robot in teleop mode or null if
-	 *        the robot is in autonomous mode.
-	 * @return FSM state for the next iteration
-	 */
-	private FSMState nextState(TeleopInput input) {
-		switch (currentState) {
-			case START_STATE:
-				if (input != null) {
-					return FSMState.OTHER_STATE;
-				} else {
-					return FSMState.START_STATE;
-				}
+        currentState = nextState(input);
+    }
+    
+    /* ======================== Private methods ======================== */
+    /**
+        * Decide the next state to transition to. This is a function of the inputs
+        * and the current state of this FSM. This method should not have any side
+        * effects on outputs. In other words, this method should only read or get
+        * values to decide what state to go to.
+        * @param input Global TeleopInput if robot in teleop mode or null if
+        *        the robot is in autonomous mode.
+        * @return FSM state for the next iteration
+        */
+    private FSMState nextState(TeleopInput input) {
+        System.out.println(currentState);
 
-			case OTHER_STATE:
-				return FSMState.OTHER_STATE;
+        if(input != null && !input.isRampToggleButtonPressed())
+            canToggleRamp = true;
 
-			default:
-				throw new IllegalStateException("Invalid state: " + currentState.toString());
-		}
-	}
+        switch (currentState) {
+            case RETRACTED:
+                if (input != null) {
+                    if(input.isShooterButtonPressed())
+                        return FSMState.SHOOT_RETRACTED;
+                    else if(input.isIntakeButtonPressed())
+                        return FSMState.INTAKING_RETRACTED;
+                    else if(input.isRampToggleButtonPressed() && canToggleRamp){
```

Comment:
I assume this is to actuate once per button press? You might be able to use the getRawButtonPressed API to do this for you.
https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/GenericHID.html#getRawButtonPressed(int)

Reply:
Yes that is correct, just pushed a version that removes that redundancy